File: app/src/main/java/com/nfcalarmclock/mediapicker/NacMediaActivity.java
package com.nfcalarmclock.mediapicker;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentOnAttachListener;
import androidx.fragment.app.FragmentManager;
import androidx.viewpager2.widget.ViewPager2;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayout.Tab;
import com.google.android.material.tabs.TabLayoutMediator;

import com.nfcalarmclock.R;
import com.nfcalarmclock.alarm.NacAlarm;
import com.nfcalarmclock.file.browser.NacFileBrowser;
import com.nfcalarmclock.media.NacMedia;
import com.nfcalarmclock.mediapicker.music.NacMusicFragment;
import com.nfcalarmclock.mediapicker.ringtone.NacRingtoneFragment;
import com.nfcalarmclock.permission.readmediaaudio.NacReadMediaAudioPermission;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.shared.NacSharedDefaults;
import com.nfcalarmclock.shared.NacSharedPreferences;
import com.nfcalarmclock.system.NacIntent;

/**
 */
public class NacMediaActivity
	extends FragmentActivity
	implements FragmentOnAttachListener,
		TabLayout.OnTabSelectedListener,
		ActivityCompat.OnRequestPermissionsResultCallback
{

	/**
	 * View pager.
	 */
	private ViewPager2 mViewPager;

	/**
	 * Adapter for the view pager.
	 */
	private NacPagerAdapter mAdapter;

	/**
	 * Tab layout.
	 */
	private TabLayout mTabLayout;

	/**
	 * List of fragments.
	 */
	private final Fragment[] mFragments;

	/**
	 * Alarm.
	 */
	private NacAlarm mAlarm;

	/**
	 * Media path.
	 */
	private String mMediaPath;

	/**
	 * Current tab position.
	 */
	private int mPosition;

	/**
	 * The tab titles.
	 */
	private final String[] mTitles = new String[2];

	/**
	 */
	public NacMediaActivity()
	{
		super();

		// Setup the fragment list
		int length = this.getTitles().length;
		this.mFragments = new Fragment[length];

		// Add a listener for when a fragment is attached
		getSupportFragmentManager().addFragmentOnAttachListener(this);
	}

	/**
	 * Select a fragment.
	 */
	private void fragmentSelected(Fragment selectedFragment)
	{
		int position = this.getPosition();
		Fragment tabFragment = this.getFragments()[position];

		// The selected fragment is undefined or does not match the position of
		// the fragment that the user should be on
		if ((selectedFragment == null) || (selectedFragment != tabFragment))
		{
			return;
		}

		// The user is on the music fragment (where they can select from their
		// own music to set as the media for an alarm), but the app has not
		// been given the permission to read audio files on the phone
		if ((selectedFragment instanceof NacMusicFragment)
				&& !NacReadMediaAudioPermission.hasPermission(this))
		{
			// Request permission to read audio files
			// TODO: Stack overflow error here
			NacReadMediaAudioPermission.requestPermission(this, NacMusicFragment.READ_REQUEST_CODE);
		}

		// Select the fragment
		((NacMediaFragment)selectedFragment).onSelected();
	}

	/**
	 * @return The alarm.
	 */
	private NacAlarm getAlarm()
	{
		return this.mAlarm;
	}

	/**
	 * @return The list of fragments.
	 */
	private Fragment[] getFragments()
	{
		return this.mFragments;
	}

	/**
	 * @return The view page adapter.
	 */
	private NacPagerAdapter getPagerAdapter()
	{
		return this.mAdapter;
	}

	/**
	 * @return The current tab position.
	 */
	private int getPosition()
	{
		return this.mPosition;
	}

	/**
	 * @return The media path.
	 */
	private String getMedia()
	{
		return this.mMediaPath;
	}

	/**
	 * @return The media type.
	 */
	private int getMediaType()
	{
		NacAlarm alarm = this.getAlarm();
		String media = this.getMedia();

		if (alarm != null)
		{
			return alarm.getMediaType();
		}
		else if (media != null)
		{
			return NacMedia.getType(this, media);
		}
		else
		{
			return NacMedia.TYPE_NONE;
		}
	}

	/**
	 * @return The tab layout.
	 */
	private TabLayout getTabLayout()
	{
		return this.mTabLayout;
	}

	/**
	 * @return The array of titles.
	 */
	private String[] getTitles()
	{
		return this.mTitles;
	}

	/**
	 * @return The view pager.
	 */
	private ViewPager2 getViewPager()
	{
		return this.mViewPager;
	}

	/**
	 */
	@Override
	public void onAttachFragment(@NonNull FragmentManager manager,
		@NonNull Fragment fragment)
	{
		Fragment[] list = this.getFragments();

		//if (list == null)
		//{
		//	this.mFragments = new Fragment[this.mTitles.length];
		//	list = this.mFragments;
		//}

		if (fragment instanceof NacMusicFragment)
		{
			list[0] = fragment;
		}
		else if (fragment instanceof NacRingtoneFragment)
		{
			list[1] = fragment;
		}
		//else if (fragment instanceof NacSpotifyFragment)
		//{
		//	list[2] = fragment;
		//}

		this.fragmentSelected(fragment);
	}

	/**
	 */
	@Override
	public void onBackPressed()
	{
		// Get the current position and the music fragment
		int position = this.getPosition();
		NacMusicFragment musicFragment = (NacMusicFragment)
			this.getFragments()[0];

		// Check if at position 0 and that the music fragment is defined. The
		// music fragment (where the user can browse for music to play for
		// an alarm, instead of a ringtone) is at position 0. Do custom stuff
		// if back is pressed for this fragment
		if ((position == 0) && (musicFragment != null))
		{
			NacFileBrowser fileBrowser = musicFragment.getFileBrowser();

			// Browser is undefined. Do nothing
			if (fileBrowser == null)
			{
				return;
			}

			// Go to previous directory if not already at root level,
			// otherwise, do a normal back press which will exit the activity
			if (!fileBrowser.isAtRoot())
			{
				fileBrowser.previousDirectory();
				return;
			}
		}

		// Normal back press which should exit the activity
		super.onBackPressed();
	}

	/**
	 */
	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setContentView(R.layout.act_sound);

		Intent intent = getIntent();
		NacSharedConstants cons = new NacSharedConstants(this);

		this.mAlarm = NacIntent.getAlarm(intent);
		this.mMediaPath = NacIntent.getMedia(intent);
		this.mViewPager = findViewById(R.id.act_sound);
		this.mTabLayout = findViewById(R.id.tab_layout);
		this.mTitles[0] = cons.getActionBrowse();
		this.mTitles[1] = cons.getAudioSources().get(3);
		this.mAdapter = new NacPagerAdapter(this);
		this.mPosition = 0;

		int mediaType = this.getMediaType();

		this.setupViewPager();
		this.setupTabColors();
		this.selectTabByMediaType(mediaType);
	}

	/**
	 */
	@Override
	public void onRequestPermissionsResult(int requestCode,
		@NonNull String[] permissions, @NonNull int[] grantResults)
	{
		super.onRequestPermissionsResult(requestCode, permissions, grantResults);

		// Request code does not match the one that this activity sent
		if (requestCode != NacMusicFragment.READ_REQUEST_CODE)
		{
			return;
		}

		// Get the music fragment
		Fragment fragment = this.getFragments()[0];

		// Permission was granted
		if ((fragment != null) && (grantResults.length > 0)
			&& (grantResults[0] == PackageManager.PERMISSION_GRANTED))
		{

			// Create a new pager adapter in order to refresh the fragments
			// in the view pager. Not sure if this is the way to do it, but
			// this is the only way that I found works
			ViewPager2 viewPager = this.getViewPager();
			NacPagerAdapter adapter = new NacPagerAdapter(this);

			// Set pager adapter on the view pager
			viewPager.setAdapter(adapter);

			// Set the new pager adapter member variable
			this.mAdapter = adapter;
		}
		// Either permission was not granted, or the fragment was null for some
		// reason. Whatever the case, just show the ringtone fragment instead
		else
		{
			this.selectTabByIndex(1);
		}
	}

	/**
	 */
	@Override
	public void onTabReselected(Tab tab)
	{
	}

	/**
	 */
	@Override
	public void onTabSelected(Tab tab)
	{
		int position = tab.getPosition();
		Fragment fragment = this.getFragments()[position];
		this.mPosition = position;

		this.fragmentSelected(fragment);
	}

	/**
	 */
	@Override
	public void onTabUnselected(Tab tab)
	{
	}

	/**
	 * Select the tab at the given index.
	 *
	 * @param  index  The index of the tab.
	 */
	private void selectTabByIndex(int index)
	{
		TabLayout tabLayout = this.getTabLayout();

		if (tabLayout != null)
		{
			Tab tab = tabLayout.getTabAt(index);

			if (tab != null)
			{
				tab.select();
				onTabSelected(tab);
			}
		}
	}

	/**
	 * Select the tab that the fragment activity should start on.
	 *
	 * @param  mediaType  The media type for when the alarm goes off.
	 */
	private void selectTabByMediaType(int mediaType)
	{
		if (NacMedia.isNone(mediaType))
		{
			this.selectTabByIndex(1);
		}
		else if (NacMedia.isFile(mediaType) || NacMedia.isDirectory(mediaType))
		{
			this.selectTabByIndex(0);
		}
		else if (NacMedia.isRingtone(mediaType))
		{
			this.selectTabByIndex(1);
		}
		else if (NacMedia.isSpotify(mediaType))
		{
			this.selectTabByIndex(2);
		}
	}

	/**
	 * Setup the tab colors.
	 */
	private void setupTabColors()
	{
		NacSharedPreferences shared = new NacSharedPreferences(this);
		NacSharedDefaults defaults = shared.getDefaults();
		TabLayout tabLayout = this.getTabLayout();

		tabLayout.setSelectedTabIndicatorColor(shared.getThemeColor());
		tabLayout.setTabTextColors(defaults.getColor(),
			shared.getThemeColor());
	}

	/**
	 * Setup the view pager.
	 */
	@SuppressWarnings("deprecation")
	private void setupViewPager()
	{
		ViewPager2 viewPager = this.getViewPager();
		NacPagerAdapter adapter = this.getPagerAdapter();
		TabLayout tabLayout = this.getTabLayout();

		// Set the pager adapter
		viewPager.setAdapter(adapter);

		// Set the tab layout mediator
		new TabLayoutMediator(tabLayout, viewPager, (tab, position) ->
			tab.setText(this.getTitles()[position])).attach();

		// Set the tab selected listener based on the Android API level
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
		{
			tabLayout.addOnTabSelectedListener(this);
		}
		else
		{
			tabLayout.setOnTabSelectedListener(this);
		}
	}

	/**
	 */
	//public static class NacPagerAdapter
	public class NacPagerAdapter
		extends FragmentStateAdapter
	{

		///**
		// * Alarm.
		// */
		//private final NacAlarm mAlarm;

		///**
		// * Tab titles.
		// */
		//private final String[] mTitles;

		/**
		 */
		public NacPagerAdapter(FragmentActivity fa)
		{
			super(fa);
		}

		/**
		 */
		//@NonNull
        @Override
		public Fragment createFragment(int position)
		{
			NacAlarm alarm = getAlarm();
			String media = getMedia();

			if (position == 0)
			{
				if (alarm != null)
				{
					return NacMusicFragment.newInstance(alarm);
				}
				else if (media != null)
				{
					return NacMusicFragment.newInstance(media);
				}
			}
			else if (position == 1)
			{
				if (alarm != null)
				{
					return NacRingtoneFragment.newInstance(alarm);
				}
				else if (media != null)
				{
					return NacRingtoneFragment.newInstance(media);
				}
			}
			//else if (position == 2)
			//{
			//	if (alarm != null)
			//	{
			//		return NacSpotifyFragment.newInstance(alarm);
			//	}
			//	else if (media!= null)
			//	{
			//		return NacSpotifyFragment.newInstance(media);
			//	}
			//}

			return null;
		}

		/**
		 * Get the number of items to swipe through.
		 *
		 * @return The number of items to swipe through.
		 */
		@Override
		public int getItemCount()
		{
			return getTitles().length;
		}

	}

}

File: app/src/main/java/com/nfcalarmclock/mediapicker/NacMediaActivity.kt
package com.nfcalarmclock.mediapicker

import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import androidx.core.app.ActivityCompat.OnRequestPermissionsResultCallback
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentOnAttachListener
import androidx.media3.common.util.UnstableApi
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.google.android.material.tabs.TabLayoutMediator
import com.nfcalarmclock.R
import com.nfcalarmclock.alarm.db.NacAlarm
import com.nfcalarmclock.media.NacMedia
import com.nfcalarmclock.mediapicker.music.NacMusicFragment
import com.nfcalarmclock.mediapicker.ringtone.NacRingtoneFragment
import com.nfcalarmclock.permission.readmediaaudio.NacReadMediaAudioPermission.hasPermission
import com.nfcalarmclock.permission.readmediaaudio.NacReadMediaAudioPermission.requestPermission
import com.nfcalarmclock.shared.NacSharedPreferences
import com.nfcalarmclock.util.NacIntent.getAlarm
import com.nfcalarmclock.util.NacIntent.getMedia
import dagger.hilt.android.AndroidEntryPoint

/**
 * Media activity.
 */
@AndroidEntryPoint
class NacMediaActivity
	: FragmentActivity(),
	FragmentOnAttachListener,
	OnTabSelectedListener,
	OnRequestPermissionsResultCallback
{

	/**
	 * View pager.
	 */
	private var viewPager: ViewPager2? = null

	/**
	 * Adapter for the view pager.
	 */
	private var pagerAdapter: NacPagerAdapter? = null

	/**
	 * Tab layout.
	 */
	private var tabLayout: TabLayout? = null

	/**
	 * List of fragments.
	 */
	private val fragments: Array<Fragment?>

	/**
	 * Alarm.
	 */
	private var alarm: NacAlarm? = null

	/**
	 * Media path.
	 */
	private var media: String? = null

	/**
	 * Current tab position.
	 */
	private var position = 0

	/**
	 * The tab titles.
	 */
	private val titles = arrayOfNulls<String>(2)

	/**
	 * The media type.
	 */
	private val mediaType: Int
		get()
		{
			return alarm?.mediaType
				?: if (media != null)
					{
						NacMedia.getType(this, media!!)
					}
					else
					{
						NacMedia.TYPE_NONE
					}
		}

	/**
	 * Constructor
	 */
	init
	{
		// Setup the fragment list
		fragments = arrayOfNulls(titles.size)

		// Add a listener for when a fragment is attached
		supportFragmentManager.addFragmentOnAttachListener(this)
	}

	/**
	 * Select a fragment.
	 */
	private fun fragmentSelected(selectedFragment: Fragment?)
	{
		val tabFragment = fragments[position]

		// The selected fragment is undefined or does not match the position of
		// the fragment that the user should be on
		if (selectedFragment == null || selectedFragment !== tabFragment)
		{
			return
		}

		// The user is on the music fragment (where they can select from their
		// own music to set as the media for an alarm), but the app has not
		// been given the permission to read audio files on the phone
		if (selectedFragment is NacMusicFragment
			&& !hasPermission(this))
		{
			// Request permission to read audio files
			// TODO: Stack overflow error here
			requestPermission(this, NacMusicFragment.READ_REQUEST_CODE)
		}

		// Select the fragment
		(selectedFragment as NacMediaFragment).onSelected()
	}

	/**
	 * Called when the fragment is attached.
	 */
	override fun onAttachFragment(manager: FragmentManager,
		frag: Fragment)
	{
		// Music fragment
		if (frag is NacMusicFragment)
		{
			fragments[0] = frag
		}
		// Ringtone fragment
		else if (frag is NacRingtoneFragment)
		{
			fragments[1] = frag
		}

		// Select the gragment
		fragmentSelected(frag)
	}

	/**
	 * Called when the back button is pressed
	 */
	override fun onBackPressed()
	{
		// Get the current position and the music fragment
		val musicFragment = fragments[0] as NacMusicFragment?

		// Check if at position 0 and that the music fragment is defined. The
		// music fragment (where the user can browse for music to play for
		// an alarm, instead of a ringtone) is at position 0. Do custom stuff
		// if back is pressed for this fragment
		if (position == 0 && musicFragment != null)
		{
			val fileBrowser = musicFragment.fileBrowser ?: return

			// Browser is undefined. Do nothing

			// Go to previous directory if not already at root level,
			// otherwise, do a normal back press which will exit the activity
			if (!fileBrowser.isAtRoot)
			{
				fileBrowser.previousDirectory()
				return
			}
		}

		// Normal back press which should exit the activity
		super.onBackPressed()
	}

	/**
	 * Called when the activity is created.
	 */
	override fun onCreate(savedInstanceState: Bundle?)
	{
		// Super
		super.onCreate(savedInstanceState)

		// Set the layout
		setContentView(R.layout.act_sound)

		// Get all the audio sources
		val audioSources = resources.getStringArray(R.array.audio_sources)

		// Set the member variables
		alarm = getAlarm(intent)
		media = getMedia(intent)
		viewPager = findViewById(R.id.act_sound)
		tabLayout = findViewById(R.id.tab_layout)
		titles[0] = getString(R.string.action_browse)
		titles[1] = audioSources[3]
		pagerAdapter = NacPagerAdapter(this)
		position = 0

		// Setup the views
		setupViewPager()
		setupTabColors()
		selectTabByMediaType(mediaType)
	}

	/**
	 * Called when the permission request receives a result.
	 */
	override fun onRequestPermissionsResult(requestCode: Int,
		permissions: Array<String>, grantResults: IntArray)
	{
		// Super
		super.onRequestPermissionsResult(requestCode, permissions, grantResults)

		// Request code does not match the one that this activity sent
		if (requestCode != NacMusicFragment.READ_REQUEST_CODE)
		{
			return
		}

		// Get the music fragment
		val musicFragment = fragments[0]

		// Permission was granted
		if (musicFragment != null && grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)
		{

			// Create a new pager adapter in order to refresh the fragments
			// in the view pager. Not sure if this is the way to do it, but
			// this is the only way that I found works
			val adapter = NacPagerAdapter(this)

			// Set pager adapter on the view pager
			viewPager!!.adapter = adapter

			// Set the new pager adapter member variable
			pagerAdapter = adapter
		}
		// Permission was denied
		else
		{
			selectTabByIndex(1)
		}
	}

	/**
	 * Called when a tab is reselected.
	 */
	override fun onTabReselected(tab: TabLayout.Tab)
	{
	}

	/**
	 * Called when a tab is selected.
	 */
	override fun onTabSelected(tab: TabLayout.Tab)
	{
		// Set the current position
		position = tab.position

		// Get the fragment at that position
		val fragment = fragments[position]

		// Select the fragment
		fragmentSelected(fragment)
	}

	/**
	 * Called when a tab is unselected.
	 */
	override fun onTabUnselected(tab: TabLayout.Tab)
	{
	}

	/**
	 * Select the tab at the given index.
	 *
	 * @param  index  The index of the tab.
	 */
	private fun selectTabByIndex(index: Int)
	{
		// Check if the tab layout is not defined yet
		if (tabLayout == null)
		{
			return
		}

		// Get the tab at the index
		val tab = tabLayout!!.getTabAt(index)

		// Check if the tab is not null
		if (tab != null)
		{
			// Select the tab
			tab.select()
			onTabSelected(tab)
		}
	}

	/**
	 * Select the tab that the fragment activity should start on.
	 *
	 * @param  mediaType  The media type for when the alarm goes off.
	 */
	private fun selectTabByMediaType(mediaType: Int)
	{
		// None
		if (NacMedia.isNone(mediaType))
		{
			selectTabByIndex(1)
		}
		// File or directory
		else if (NacMedia.isFile(mediaType) || NacMedia.isDirectory(mediaType))
		{
			selectTabByIndex(0)
		}
		// Ringtone
		else if (NacMedia.isRingtone(mediaType))
		{
			selectTabByIndex(1)
		}
		// Spotify
		//else if (NacMedia.isSpotify(mediaType))
		//{
		//	selectTabByIndex(2)
		//}
	}

	/**
	 * Setup the tab colors.
	 */
	private fun setupTabColors()
	{
		val shared = NacSharedPreferences(this)
		val defaultColor = resources.getInteger(R.integer.default_color)

		// Set the color
		tabLayout!!.setSelectedTabIndicatorColor(shared.themeColor)
		tabLayout!!.setTabTextColors(defaultColor, shared.themeColor)
	}

	/**
	 * Setup the view pager.
	 */
	@Suppress("deprecation")
	private fun setupViewPager()
	{
		// Set the pager adapter
		viewPager!!.adapter = pagerAdapter

		// Set the tab layout mediator
		TabLayoutMediator(tabLayout!!, viewPager!!) { tab: TabLayout.Tab, position: Int ->
			tab.text = titles[position]
		}.attach()

		// Set the tab selected listener based on the Android API level
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
		{
			tabLayout!!.addOnTabSelectedListener(this)
		}
		// API < 26
		else
		{
			tabLayout!!.setOnTabSelectedListener(this)
		}
	}

	/**
	 * Pager adapter.
	 */
	inner class NacPagerAdapter(fa: FragmentActivity)
		: FragmentStateAdapter(fa)
	{

		/**
		 * Create the fragment.
		 */
		@UnstableApi
		override fun createFragment(position: Int): Fragment
		{
			return when (position)
			{

				// Music fragment
				0 ->
				{
					createMusicFragment()
				}

				// Ringtone fragment
				1 ->
				{
					createRingtoneFragment()
				}

				// Unknown
				else ->
				{
					createRingtoneFragment()
				}

			}

			//else if (position == 2)
			//{
			//	if (alarm != null)
			//	{
			//		return NacSpotifyFragment.newInstance(alarm);
			//	}
			//	else if (media!= null)
			//	{
			//		return NacSpotifyFragment.newInstance(media);
			//	}
			//}
		}

		/**
		 * Create a music fragment.
		 */
		private fun createMusicFragment(): Fragment
		{
			// Check that alarm is not null
			return if (alarm != null)
			{
				NacMusicFragment.newInstance(alarm)
			}
			// Check that media is not null
			else if (media != null)
			{
				NacMusicFragment.newInstance(media)
			}
			// Return an empty music fragment
			else
			{
				NacMusicFragment()
			}
		}

		/**
		 * Create a ringtone fragment.
		 */
		@UnstableApi
		private fun createRingtoneFragment(): Fragment
		{
			// Check that alarm is not null
			return if (alarm != null)
			{
				NacRingtoneFragment.newInstance(alarm)
			}
			// Check that media is not null
			else if (media != null)
			{
				NacRingtoneFragment.newInstance(media)
			}
			// Return an empty music fragment
			else
			{
				NacRingtoneFragment()
			}
		}

		/**
		 * Get the number of items to swipe through.
		 *
		 * @return The number of items to swipe through.
		 */
		override fun getItemCount(): Int
		{
			return titles.size
		}

	}

}